src/envs/pybullet_env.py

```diff
@@ -0,0 +1,423 @@
+                    physicsClientId=self._physics_client_id)
+
+        # If placing, remove the grasp constraint.
+        if self._held_constraint_id is not None and \
+            f_delta > self._finger_action_tol:
+            p.removeConstraint(self._held_constraint_id,
+                               physicsClientId=self._physics_client_id)
+            self._held_constraint_id = None
+            self._held_obj_id = None
+
+        self._current_state = self._get_state()
+        return self._current_state.copy()
+
+    def _detect_held_object(self) -> Optional[int]:
+        """Return the PyBullet object ID of the held object if one exists.
+
+        If multiple objects are within the grasp tolerance, return the
+        one that is closest.
+        """
+        expected_finger_normals = {
+            self._pybullet_robot.left_finger_id: np.array([0., 1., 0.]),
+            self._pybullet_robot.right_finger_id: np.array([0., -1., 0.]),
+        }
+        closest_held_obj = None
+        closest_held_obj_dist = float("inf")
+        for obj_id in self._get_object_ids_for_held_check():
+            for finger_id, expected_normal in expected_finger_normals.items():
+                # Find points on the object that are within grasp_tol distance
+                # of the finger. Note that we use getClosestPoints instead of
+                # getContactPoints because we still want to consider the object
+                # held even if there is a tiny distance between the fingers and
+                # the object.
+                closest_points = p.getClosestPoints(
+                    bodyA=self._pybullet_robot.robot_id,
+                    bodyB=obj_id,
+                    distance=self._grasp_tol,
+                    linkIndexA=finger_id,
+                    physicsClientId=self._physics_client_id)
+                for point in closest_points:
+                    # If the contact normal is substantially different from
+                    # the expected contact normal, this is probably an object
+                    # on the outside of the fingers, rather than the inside.
+                    # A perfect score here is 1.0 (normals are unit vectors).
+                    contact_normal = point[7]
+                    if expected_normal.dot(contact_normal) < 0.5:
+                        continue
+                    # Handle the case where multiple objects pass this check
+                    # by taking the closest one. This should be rare, but it
+                    # can happen when two objects are stacked and the robot is
+                    # unstacking the top one.
+                    contact_distance = point[8]
+                    if contact_distance < closest_held_obj_dist:
+                        closest_held_obj = obj_id
+                        closest_held_obj_dist = contact_distance
+        return closest_held_obj
+
+    def _create_block(self, block_num: int, size: float) -> int:
```

Comment:
i'm trying to reason about the purview of this base class. on one hand, we're changing `block` to `obj` in some variable names above. on the other hand, we are including this method in this class, so the class is specific to blocks, in some sense... what's our guiding philosophy?

Reply:
pulled this method out as a top-level method in pybullet_envs.py